internal/jujuclient/dial.go

```diff
@@ -76,14 +75,45 @@ func (d *Dialer) createLoginRequest(ctx context.Context, ctl *dbmodel.Controller
 	jwtString := base64.StdEncoding.EncodeToString(jwt)
 
 	return &jujuparams.LoginRequest{
-		AuthTag:       names.NewUserTag("admin").String(),
+		AuthTag:       userTag.String(),
 		ClientVersion: jujuClientVersion,
 		Token:         jwtString,
 	}, nil
 }
 
+func (d *Dialer) createAdminLoginRequest(ctx context.Context, ctl *dbmodel.Controller, modelTag names.ModelTag, additionalPermissions map[string]string) (*jujuparams.LoginRequest, error) {
+	permissions := make(map[string]string)
+	permissions[ctl.ResourceTag().String()] = "superuser"
+	if modelTag.Id() != "" {
+		permissions[modelTag.String()] = string(params.ModelAdminAccess)
+	}
+	for k, v := range additionalPermissions {
+		permissions[k] = v
+	}
+	return d.createLoginRequest1(ctx, ctl.ResourceTag(), names.NewUserTag(adminUser), permissions)
+}
+
+func (d *Dialer) createLoginRequest(ctx context.Context, ctl *dbmodel.Controller, modelTag names.ModelTag, user *openfga.User, additionalPermissions map[string]string) (*jujuparams.LoginRequest, error) {
+	p := make(map[string]string)
+	ctlRelation := user.GetControllerAccess(ctx, ctl.ResourceTag())
+	if ctlRelation == ofganames.AdministratorRelation {
+		p[ctl.ResourceTag().String()] = "superuser"
+	} else {
+		p[ctl.ResourceTag().String()] = "login"
+	}
+	modelRelation := user.GetModelAccess(ctx, modelTag)
+	if modelRelation != ofganames.NoRelation {
+		p[modelTag.String()] = permissions.ToModelAccessString(modelRelation)
+	}
+
+	for k, v := range additionalPermissions {
+		p[k] = v
+	}
+	return d.createLoginRequest1(ctx, ctl.ResourceTag(), user.ResourceTag(), p)
+}
+
 // Dial implements jimm.Dialer.
-func (d *Dialer) Dial(ctx context.Context, ctl *dbmodel.Controller, modelTag names.ModelTag, requiredPermissions map[string]string) (juju.API, error) {
+func (d *Dialer) Dial(ctx context.Context, ctl *dbmodel.Controller, modelTag names.ModelTag, user *openfga.User, withPermissions map[string]string) (juju.API, error) {
```

Comment:
Although I think the change to the Dial function's signature is fine for now. It would be good to revisit it, because it's getting too long imo.
Some ideas,
- We create `AdminDialer` and `UserDialer` structs whose `NewUserDialer` function accepts a dialer, user and permission map.
- We pass the user's permission in the context and extract them in the Dial method.

Reply:
definitely agree.. this is a bit of a mess atm.